ospfd/ospf_te.c

```diff
@@ -1574,138 +1617,1655 @@ void ospf_mpls_te_lsa_schedule(struct mpls_te_link *lp, enum lsa_opcode opcode)
+			attr.standard.te_metric = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_TE_METRIC);
+			break;
+		case TE_LINK_SUBTLV_MAX_BW:
+			memcpy(&valf, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.standard.max_bw = ntohf(valf);
+			SET_FLAG(attr.flags, LS_ATTR_MAX_BW);
+			break;
+		case TE_LINK_SUBTLV_MAX_RSV_BW:
+			memcpy(&valf, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.standard.max_rsv_bw = ntohf(valf);
+			SET_FLAG(attr.flags, LS_ATTR_MAX_RSV_BW);
+			break;
+		case TE_LINK_SUBTLV_UNRSV_BW:
+			memcpy(tabf, value, TE_LINK_SUBTLV_UNRSV_SIZE);
+			for (int i = 0; i < MAX_CLASS_TYPE; i++)
+				attr.standard.unrsv_bw[i] = ntohf(tabf[i]);
+			SET_FLAG(attr.flags, LS_ATTR_UNRSV_BW);
+			break;
+		case TE_LINK_SUBTLV_RSC_CLSCLR:
+			memcpy(&val32, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.standard.admin_group = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_ADM_GRP);
+			break;
+		case TE_LINK_SUBTLV_LLRI:
+			memcpy(tab32, value, TE_LINK_SUBTLV_LLRI_SIZE);
+			attr.standard.local_id = ntohl(tab32[0]);
+			attr.standard.remote_id = ntohl(tab32[1]);
+			SET_FLAG(attr.flags, LS_ATTR_LOCAL_ID);
+			SET_FLAG(attr.flags, LS_ATTR_NEIGH_ID);
+			break;
+		case TE_LINK_SUBTLV_RIP:
+			memcpy(&addr, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.standard.remote_addr = addr;
+			SET_FLAG(attr.flags, LS_ATTR_REMOTE_ADDR);
+			break;
+		case TE_LINK_SUBTLV_RAS:
+			memcpy(&val32, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.standard.remote_as = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_REMOTE_AS);
+			break;
+		case TE_LINK_SUBTLV_AV_DELAY:
+			memcpy(&val32, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.delay = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_DELAY);
+			break;
+		case TE_LINK_SUBTLV_MM_DELAY:
+			memcpy(tab32, value, TE_LINK_SUBTLV_MM_DELAY_SIZE);
+			attr.extended.min_delay = ntohl(tab32[0]);
+			attr.extended.max_delay = ntohl(tab32[1]);
+			SET_FLAG(attr.flags, LS_ATTR_MIN_MAX_DELAY);
+			break;
+		case TE_LINK_SUBTLV_DELAY_VAR:
+			memcpy(&val32, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.jitter = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_JITTER);
+			break;
+		case TE_LINK_SUBTLV_PKT_LOSS:
+			memcpy(&val32, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.pkt_loss = ntohl(val32);
+			SET_FLAG(attr.flags, LS_ATTR_PACKET_LOSS);
+			break;
+		case TE_LINK_SUBTLV_RES_BW:
+			memcpy(&valf, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.rsv_bw = ntohf(valf);
+			SET_FLAG(attr.flags, LS_ATTR_RSV_BW);
+			break;
+		case TE_LINK_SUBTLV_AVA_BW:
+			memcpy(&valf, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.ava_bw = ntohf(valf);
+			SET_FLAG(attr.flags, LS_ATTR_AVA_BW);
+			break;
+		case TE_LINK_SUBTLV_USE_BW:
+			memcpy(&valf, value, TE_LINK_SUBTLV_DEF_SIZE);
+			attr.extended.used_bw = ntohf(valf);
+			SET_FLAG(attr.flags, LS_ATTR_USE_BW);
+			break;
+		default:
+			break;
+		}
+		sum += TLV_SIZE(tlvh);
+	}
+
+	/* Get corresponding Edge from Link State Data Base */
+	edge = get_edge(ted, attr.adv, attr.standard.local);
```

Comment:
Check edge pointer (crashed in my tests -mixing frr and cisco- without ospfv2 mpls-te configuration)

Reply:
Can you provide more detail about the crash ? get_edge() function always return a valid pointer: if there is no edge that correspond to this IPv4 address, a new one is created. So, the only problem could happen if there is no more memory available i.e. XCALLOC failed.

Reply:
In addition, if you not enable mpls-te on cisco side, this function is never call as the Cisco is not flooding any TE Opaque LSA.

So, the failure is very strange. Can you reproduce the crash once activating the MPLS-TE log (debug ospf te) and provide the crash information + log ?

Reply:
Well, first, the code doesn't support non point-to-point interface. Result is unexpected.

Second, from what I saw on the log, the Cisco seems not sending any opaque LSA for TE nor for SR. Only router LSA have been parsed by the MPLS-TE feature.

Can you check that the FRR router received such Opaque LSA by running `sh ip ospf database` and `sh ip ospf database opaque-area`  ? In addition, you could look to the content of the TED with 'sh ip ospf mpls-te database` command.

Reply:
I confirme the crash. Vertex, Edge and Subnet key must not be equal to 0 and the Link State Library function check that. As the Edge key is based on the local IP address, it is normal that the Edge and Attributes are not correctly forge as it seems that the local IP address is not detected. It seems that the Opaque LSA TLV is empty as nothing is detected and that the length of  TLV is 0. Again, look at the LSA of the ospf database to verify that there are valid.

Reply:
After protecting the access to edge ospf_te_parse_te I can run those commands and this is the result

Reply:
I finally understand what's wrong. According to RFC3630, Router ID must be advertise only once. In FRR, we choose to insert the Router ID TLV within the same Opaque LSA as for the TLV 2 Traffic Engineering. This is certainly not good and not respect correctly the RFC.

In turn, Cisco apply the RFC by sending 2 Opaque LSA: One with the Router ID TLV = 1 and one per interface with Traffic Engineering TLV = 2. When the new code parse an Opaque LSA with the Router ID TLV = 1, it doesn't found any link information, thus failed and crash as the Edge object is not created when link address is equal to 0.

So, the correct way to handle this situation, is to detect that the Opaque LSA carry a Traffic Engineering TLV = 2 and stop processing if only Router ID TLV = 1 is present. I will update the code accordingly.

Reply:
Ah, and to look in detail to the content of an opaque LSA, you need to provide the LSA ID not the Router ID e.g.
`sh ip ospf database opaque-area 1.0.0.1` in your setup